release_util/management/commands/__init__.py

```diff
@@ -132,164 +134,215 @@ def state(self):
         return yaml.safe_dump(self.migration_state)
 
 
+def _remove_escape_characters(s):
+    ansi_escape = re.compile(r'(\x9B|\x1B\[)[0-?]*[ -\/]*[@-~]')
+    return ansi_escape.sub('', s)
+
+
+class MigrationSessionError(ValueError):
+    pass
+
+
 class MigrationSession(object):
     """
     Class which performs migrations while keeping track of the state of each migration.
     Provides the state of all migrations post-migration.
-    """
-    def __init__(self, stderr, database_name):
-        self.to_apply = []
-        self.migration_state = {
-            'success': [],
-            'failure': None,
-            'unapplied': [],
-            'rollback_commands': [],
-        }
-        self.timer = default_timer
-        self.stderr = stderr
-        self.database_name = database_name
 
-        # Execute command to find/record the unapplied migrations.
-        out = StringIO()
-        call_command("show_unapplied_migrations",
-                     interactive=False,
-                     stdout=out,
-                     database=self.database_name,
-                     standalone=False)
+    If you want to run specific migrations, you can either initialize it with
+    a list of migration tuples, or add migrations by calling add_migrations(),
+    then call apply().
 
-        # Load the passed-in YAML into a dictionary.
-        self.input_migrations = yaml.safe_load(out.getvalue())
+    If you want to run all migrations, simply call apply_all(). Calling
+    apply_all() while a MigrationSession has migrations added to it will raise
+    an error.
 
-        # Build a list of migrations that will be applied.
-        for migration in self.input_migrations['migrations']:
-            self.to_apply.append(migration)
+    Once either apply() or apply_all() has been called, calling apply(),
+    apply_all(), or add_migrations() will raise an error.
+
+    The "stderr" argument is deprecated and not used.
+    """
+    def __init__(self, stderr, database_name, migrations=None):
+        self._to_apply = []
+        self._migration_state = []
+        self._timer = default_timer
+        self._closed = False
```

Comment:
Why? Do you anticipate us having other threads wait on `_closed`? It wouldn't make it harder to change back to `True` since events have a `.clear()` method.